backend/src/apis/fetch.rs

```diff
@@ -1,151 +1,102 @@
-#[tokio::main]
-async fn fetch_balances_per_addr(
-    address: String,
-    asset_contract_addresses: Vec<String>,
-) -> Result<Vec<u64>, Box<dyn Error>> {
-    // Create a header map
-    let mut headers = HeaderMap::new();
-
-    // Load .env file
-    dotenv::dotenv().ok();
-
-    // Access API key from the environment
-    let api_key = std::env::var("COVALENT_API_KEY").unwrap();
-
-    // Add `Content-Type` header
-    headers.insert(CONTENT_TYPE, HeaderValue::from_static("application/json"));
-
-    // Add `Authorization` header
-    let auth_value = format!("Basic {}", encode(api_key));
-    headers.insert(AUTHORIZATION, HeaderValue::from_str(&auth_value)?);
-
-    // Form URL
-    let url = format!(
-        "https://api.covalenthq.com/v1/eth-mainnet/address/{}/balances_v2/",
-        address
-    );
-
-    // Send a GET request to the API
-    let res = reqwest::Client::new()
-        .get(&url)
-        .headers(headers)
-        .send()
-        .await?
-        .json::<Response>()
-        .await?;
-
-    // Get balances for the specific tokens
-    let filtered_balances =
-        filter_balances_by_token_contracts(asset_contract_addresses, &res).unwrap();
-
-    Ok(filtered_balances)
-}
-
-/// This function filters out only the balances corresponding to an asset listed in asset_contract_addresses.
-fn filter_balances_by_token_contracts(
-    asset_contract_addresses: Vec<String>,
-    response: &Response,
-) -> Result<Vec<u64>, &'static str> {
-    let mut balances = Vec::new();
-    for contract_adddress in asset_contract_addresses {
-        if let Some(item) = response.data.items.iter().find(|&item| {
-            item.contract_address.to_ascii_lowercase() == contract_adddress.to_ascii_lowercase()
-        }) {
-            match item.balance.parse::<u64>() {
-                Ok(num) => balances.push(num),
-                Err(e) => println!("Failed to parse string: {}", e),
-            }
-        } else {
-            balances.push(0 as u64);
+pub async fn fetch_asset_sums<'a, M: Middleware + 'a>(
+    client: Arc<SignerMiddleware<Provider<Http>, LocalWallet>>,
+    contracts: Vec<Box<dyn BalanceFromContract<M> + Send>>,
+    asset_addresses: Vec<H160>,
+) -> Result<Vec<U256>, Box<dyn Error>> {
+    // For checking connectivity
+    client.get_block_number().await?;
+
+    let mut result: Vec<U256> = Vec::new();
+
+    let mut get_balance_futures = Vec::new();
+    for addr in asset_addresses.clone() {
+        get_balance_futures.push(client.get_balance(addr, None));
+    }
+    let get_balances = try_join_all(get_balance_futures).await?;
+    let sum_eth_balance = get_balances
+        .into_iter()
+        .reduce(|acc, balance| acc + balance)
+        .unwrap();
+    result.push(sum_eth_balance);
+
+    // Most in case, the number of contracts is less than the number of asset addresses
+    // Iterating over contracts first is more efficient
+    let mut sum_erc_balance = U256::zero();
```

Comment:
"erc" here can be misleading for someone unfamiliar with the logic. Right now it is specifically erc20 but if you want to avoid being specific you can call it "token_balance" to distinguish with, e.g., "ether_balance".

Reply:
Updated at 9ada335